fix: #15697 Assignment warning modal poping up ,even when hosts are added .

vijayraghav-io opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes #15697 
- Fixes CAL-4005 (Linear issue number - should be visible at the bottom of the GitHub issue description)

After Changes:

https://www.loom.com/share/96a6c9c3cee543bcb716a3b01aa8306f?sid=10a77132-294d-4d3e-b25e-5a6901a40d5c

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A -  I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

vijayraghav-io committed:
removed reduntand check on assignedusers

sean-brydon reviewed (dismissed):
Hi - we also need to take into consideration that assignedUsers can be empty for MANAGED event types. 

Also please refer to the issue - we should try and extract this logic out of the useEffect into a new hook that should be testable

vijayraghav-io commented:
> Hi - we also need to take into consideration that assignedUsers can be empty for MANAGED event types.
> 
> Also please refer to the issue - we should try and extract this logic out of the useEffect into a new hook that should be testable

Yes this check is already done for ManagedEventType on line 413

sean-brydon commented:
> > Hi - we also need to take into consideration that assignedUsers can be empty for MANAGED event types.
> > Also please refer to the issue - we should try and extract this logic out of the useEffect into a new hook that should be testable
> 
> Yes this check is already done for ManagedEventType on line 413

Good spot - we also want this extracted and tested. 

Thanks

vijayraghav-io committed:
update for managed eventtype

vijayraghav-io committed:
formatted

vijayraghav-io commented:
working on extracting the logic out of useEffect() and add tests

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into fix_#15697_availabilityModal

vijayraghav-io committed:
comments

vijayraghav-io committed:
Updated comments

vijayraghav-io commented:
in progress... will update once the extraction and tests are complete

vijayraghav-io committed:
extracted checkForEmptyAssignment and added tests

vijayraghav-io committed:
Merge branch 'fix_#15697_availabilityModal' of https://github.com/vijayraghav-io/cal.com into fix_#15697_availabilityModal

vijayraghav-io commented:
@sean-brydon , extracted out the logic to check if assignments are empty and added test cases, retained the route handler in useEffect. Please let me know if this is what expected.

vijayraghav-io committed:
updated comments

anikdhabal commented on apps/web/modules/event-types/views/event-types-single-view.tsx:
```diff
@@ -401,25 +402,24 @@ const EventTypePage = (props: EventTypeSetupProps & { allActiveWorkflows?: Workf
     formState: { isDirty: isFormDirty, dirtyFields },
   } = formMethods;
 
+  const assignedUsers = formMethods.watch("children");
+  const hosts = formMethods.watch("hosts");
+  const assignAllTeamMembers = formMethods.watch("assignAllTeamMembers");
+
```
As we check values using the form watch method, the modal isn't shown if hosts or users are assigned, but the form isn't saved. Please fix this.

vijayraghav-io replied:
ack

vijayraghav-io replied:
I had thought since we are allowing user to navigate away from the form without saving, so this should be ok.
Anyway, updated to handle this scenario also.

vijayraghav-io committed:
updated to show dialog also when not yet saved

anikdhabal reviewed (commented):
@vijayraghav-io pls fix the conflicts

vijayraghav-io commented:
> @vijayraghav-io pls fix the conflicts

sure

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into fix_#15697_availabilityModal

vijayraghav-io committed:
update translations

vijayraghav-io commented:
@anikdhabal , done

anikdhabal commented on apps/web/components/eventtype/AssignmentWarningDialog.tsx:
```diff
@@ -39,7 +39,7 @@ const AssignmentWarningDialog = (props: AssignmentWarningDialogProps) => {
             }}
             color="minimal">
-            {t("go_back_and_assign")}
+            {t("go_back_assign_and_save")}
```
I think we don't need to change it. In our app, we need to save maximum places in order to update, so it was understandable.

vijayraghav-io replied:
Oh, ok!

anikdhabal commented on apps/web/modules/event-types/views/event-types-single-view.tsx:
```diff
@@ -400,40 +402,50 @@ const EventTypePage = (props: EventTypeSetupProps & { allActiveWorkflows?: Workf
+  const hosts = formMethods.watch("hosts");
+  const assignAllTeamMembers = formMethods.watch("assignAllTeamMembers");
+
+  useEffect(() => {
```
As per issue we should try and extract this logic out of the useEffect into a new hook

vijayraghav-io replied:
watch was used to make form dynamic, since we rely on Saved State only, updated.

vijayraghav-io committed:
updated comments

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into fix_#15697_availabilityModal

vijayraghav-io committed:
updated translation

anikdhabal commented on apps/web/lib/checkForEmptyAssignment.ts:
```diff
@@ -0,0 +1,28 @@
+  assignAllTeamMembers,
+  isTeamEvent,
+}: {
+  assignedUsers: any;
+  hosts: any;
```
This is not good practice and would be better to avoid it. Instead, you can import the types `HOST` and `ChildrenEventType` from the main event-type component or declare them here.

vijayraghav-io replied:
updated

vijayraghav-io committed:
updated types

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into fix_#15697_availabilityModal

anikdhabal commented on apps/web/lib/checkForEmptyAssignment.ts:
```diff
@@ -0,0 +1,30 @@
+    isTeamEvent &&
+    // For managed eventtype check if assigned users are empty.
+    // For non-managed eventtype check if hosts are empty.
+    (isManagedEventType ? assignedUsers.length === 0 : hosts.length === 0)
```
Hey as @sean-brydon suggested that assignedUsers can be empty for ManagedEventType. So pls fix it

vijayraghav-io replied:
No, he is trying to say to consider the scenario of ManagedEventType having empty assigned users.
Also , If we refer to current main branch code :
https://github.com/calcom/cal.com/blob/33e1abf4a7324b3bd2c9a0d5708379e51636234b/apps/web/modules/event-types/views/event-types-single-view.tsx#L412

Here warning dialog is shown if ManagedEventType and assignedusers are empty.

anikdhabal replied:
Ahh, got it

anikdhabal commented on apps/web/modules/event-types/views/event-types-single-view.tsx:
```diff
@@ -144,6 +146,8 @@ const querySchema = z.object({
 export type EventTypeSetupProps = RouterOutputs["viewer"]["eventTypes"]["get"];
 export type EventTypeSetup = RouterOutputs["viewer"]["eventTypes"]["get"]["eventType"];
+export type EventTypeAssignedUsers = RouterOutputs["viewer"]["eventTypes"]["get"]["eventType"]["children"];
+export type EventTypeHosts = RouterOutputs["viewer"]["eventTypes"]["get"]["eventType"]["hosts"];
```
"No need to declare it again; the type `Host` is already present in file. Please export it."

vijayraghav-io replied:
Yes , i tried using type - Host already defined in event-types-single-view.tsx.
Since the prop-priority is non-null, it is breaking other components.

anikdhabal commented on apps/web/lib/checkForEmptyAssignment.ts:
```diff
@@ -0,0 +1,30 @@
+  if (isTeamEvent && assignAllTeamMembers) {
+    return false;
+  } else if (
+    isTeamEvent &&
```
No need to add `isTeamEvent` check here.

vijayraghav-io replied:
Updated with more comments.

anikdhabal commented on apps/web/lib/checkForEmptyAssignment.ts:
```diff
@@ -0,0 +1,30 @@
+  assignAllTeamMembers: boolean;
+  isTeamEvent: boolean;
+}): boolean {
+  if (isTeamEvent && assignAllTeamMembers) {
```
`isTeamEvent` should be an independent check and should be the first check there. No need to reuse it.

vijayraghav-io replied:
Updated

anikdhabal commented on apps/web/test/lib/checkForEmptyAssignment.test.ts:
Need to update the test case also for managed event type.

anikdhabal reviewed (changes requested):
Left some blocking comments. Pls address them

vijayraghav-io committed:
chore: update naming and comments

anikdhabal commented:
@vijayraghav-io I think you committed two files mistakenly. And the component or file name is correct, no need to change.

vijayraghav-io committed:
Delete apps/web/lib/checkForEmptyAssignment.ts

deleted since file was renamed

vijayraghav-io committed:
Delete apps/web/test/lib/checkForEmptyAssignment.test.ts

deleted as the file is renamed to be more descriptive

anikdhabal commented on apps/web/lib/hasEmptyAssignment.ts:
```diff
@@ -0,0 +1,43 @@
+  isTeamEvent: boolean;
+}): boolean {
+  // if Non-Team event, no need to check for assignments.
+  if (!isTeamEvent) {
+    return false;
+  }
```
@vijayraghav-io I suggested you make this independent test in the `handleRouteChange` function not here. Should be present above this `!leaveWithoutAssigningHosts.current &&`

vijayraghav-io committed:
update to do teamcheck before

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into fix_#15697_availabilityModal

anikdhabal commented:
@vijayraghav-io pls revert the name back to `checkForEmptyAssignment` or something like `isAssignmentIncomplete`. Currently it does not feel good.

vijayraghav-io committed:
chore: revert name to CheckForEmptyAssignment

anikdhabal commented on apps/web/lib/checkForEmptyAssignment.ts:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  // For Managed EventType, check if assigned users are empty.
+  if (schedulingType === "MANAGED" && assignedUsers.length === 0) {
+    return true;
+  }
+
+  // For Non-Managed EventType (i.e RoundRobin or Collective), check if hosts are empty.
+  if ((schedulingType === "ROUND_ROBIN" || schedulingType === "COLLECTIVE") && hosts.length === 0) {
+    return true;
+  }
```
```
if (isManagedEventType ? assignedUsers.length === 0 : hosts.length === 0)) {
    return true;
  }
```
This simply remove all the extra work and import. I donno why you changed it

vijayraghav-io replied:
Both are doing exactly the same, i thought this will be more clearer and readable / understandable, because there was some confusion with ManagedEventType earlier.

vijayraghav-io replied:
Its the same earlier the check if SchedulingType === 'MANAGED' was done before calling the function, now SchedulingType is passed as argument. Also if new SchedulingType is added in future, we can still handle this inside this fn only

anikdhabal replied:
Yes, I know. However, it would be simpler. If we add the new SchedulingType in the future; we would need to update not only here but also in most of the places. PLs do it.

anikdhabal commented on apps/web/modules/event-types/views/event-types-single-view.tsx:
```diff
@@ -400,40 +403,35 @@ const EventTypePage = (props: EventTypeSetupProps & { allActiveWorkflows?: Workf
+    const handleRouteChange = (url: string) => {
+      const paths = url.split("/");
+
+      if (
+        team &&
```
change it to `!!team`

anikdhabal commented:
@vijayraghav-io Left a few suggestions, please address them, and finally test all the conditions locally to ensure that everything is handled.

vijayraghav-io committed:
updated check conditions

anikdhabal commented:
@vijayraghav-io can you provide a loom video of the result. Make sure to test all the event types.

vijayraghav-io committed:
updated useEffect dependencies

vijayraghav-io commented:
@anikdhabal , just updated dependencies array for useEffect(), the edgecase (mentioned in prev comment) is also covered.

Tested once again for all EventTypes (RR, Collective, Managed) , working fine for all scenarios (including the edge case). Here is the video for same. Please check.

https://www.loom.com/share/96a6c9c3cee543bcb716a3b01aa8306f?sid=c4e8dc06-abe6-419f-9394-09ae4be08d8a

All Good to go!

anikdhabal commented:
@vijayraghav-io good observation and great works. Tested it, works fine now.

anikdhabal reviewed (approved):
LGTM! Great works @vijayraghav-io

anikdhabal committed:
Merge branch 'main' into fix_#15697_availabilityModal

anikdhabal committed:
Merge branch 'main' into fix_#15697_availabilityModal
